add docs for subscription APIs (#1698)
okta/okta-developer-docs: packages/@okta/vuepress-site/docs/reference/api/roles/index.md

HTTP/1.1 204 No Content
```

> ### Subscriptions Operations

Author: @jakubvul-okta Let me know your idea about it 😄

        "channels": [
            "email"
        ],
        "status": "subscribed",
        "_links": {
            "unsubscribe": {
                "href": "http://rain.okta1.com:1802/api/v1/roles/SUPER_ADMIN/subscriptions/REPORT_SUSPICIOUS_ACTIVITY/unsubscribe",
                "hints": {
                    "allow": [
                        "POST"
                    ]
                }
            },
            "self": {
                "href": "http://rain.okta1.com:1802/api/v1/roles/SUPER_ADMIN/subscriptions/REPORT_SUSPICIOUS_ACTIVITY"
            }
        }
    }
]
```

#### Get subscription of a Role with a specific type

<ApiLifecycle access="ea" />

<ApiOperation method="get" url="/api/v1/roles/${roleType}/subscriptions/${notificationType}" />

Get subscriptions of a Role with a specific type

##### Request parameters

| Parameter        | Description                                            | Param Type  | DataType                                    | Required  |
| :--------------- | :----------------------------------------------------- | :---------- | :------------------------------------------ | :-------- |

> | roleType         | `type` of a Role                                       | URL         | [Role Type](#role-types)                    | TRUE      |

Reviewer: some of the spacing in different tables seem off. I know that's not how it will be rendered but still it'll be nice to have rows/columns of the table use equal spacing

Author: It looked bad in git but much better in intelliJ with wider screen 😃

| notificationType | `type` of a notification                               | URL         | [Notification Type](#notification-type)    | TRUE      |

##### Response parameters

[Subscription](#subscription-object)

##### Request example

```bash
curl -v -X GET \
-H "Accept: application/json" \
-H "Content-Type: application/json" \
-H "Authorization: SSWS ${api_token}" \
"https://${yourOktaDomain}/api/v1/roles/SUPER_ADMIN/subscriptions/IWA_AGENT"
```

##### Response example

```json
{
    "notificationType": "IWA_AGENT",
    "channels": [
        "email"
    ],
    "status": "subscribed",
    "_links": {
        "unsubscribe": {
            "href": "http://rain.okta1.com:1802/api/v1/roles/SUPER_ADMIN/subscriptions/IWA_AGENT/unsubscribe",
            "hints": {
                "allow": [
                    "POST"
                ]
            }
        },
        "self": {
            "href": "http://rain.okta1.com:1802/api/v1/roles/SUPER_ADMIN/subscriptions/IWA_AGENT"
        }
    }
}
```

#### Subscribe to a specific type

<ApiLifecycle access="ea" />

<ApiOperation method="post" url="/api/v1/roles/${roleType}/subscriptions/${notificationType}/subscribe" />

> Make a Role subscribes to a specific type

Author: ApiValidationException will be thrown

> Make a Role subscribes to a specific type. When you change the subscription status of a role, it will override the subscription of any individual user of that role.

Reviewer 2: ```suggestion
Subscribes a Role to a specific type. When you change the subscription status of a role, it will override the subscription of any individual user of that role.
```

##### Request parameters

| Parameter        | Description                                            | Param Type  | DataType                                 | Required  |
| :--------------- | :----------------------------------------------------- | :---------- | :--------------------------------------- | :-------- |
| roleType         | `type` of a Role                                       | URL         | [Role Type](#role-types)                 | TRUE      |
| notificationType | `type` of a notification                               | URL         | [Notification Type](#notification-type) | TRUE      |

##### Response parameters

``` http
HTTP/1.1 200 OK
```

##### Request example

```bash
curl -v -X POST \
-H "Accept: application/json" \
-H "Content-Type: application/json" \
-H "Authorization: SSWS ${api_token}" \
"https://${yourOktaDomain}/api/v1/roles/SUPER_ADMIN/subscriptions/IWA_AGENT/subscribe"
```

##### Response example

``` http
HTTP/1.1 200 OK
```

#### Unsubscribe to a specific type

<ApiLifecycle access="ea" />

> <ApiOperation method="post" url="/api/v1/roles/${roleType}/subscriptions/${notificationType}/unsubscribe" />

Reviewer: something which I think we clarify in our product help docs is that every time a role-wide subscription changes, it results in overriding individual subscriptions. It may be worth adding as a note here as well: https://help.okta.com/en/prod/Content/Topics/Settings/set_default_notifications.htm

> Make a Role unsubscribes to a specific type. When you change the subscription status of a role, it will override the subscription of any individual user of that role.

Reviewer 2: ```suggestion
Unsubscribes a Role from a specific type. When you change the subscription status of a role, it will override the subscription of any individual user of that role.
```

##### Request parameters

| Parameter        | Description                                            | Param Type  | DataType                                 | Required  |
| :--------------- | :----------------------------------------------------- | :---------- | :--------------------------------------- | :-------- |
| roleType         | `type` of a Role                                       | URL         | [Role Type](#role-types)                 | TRUE      |
| notificationType | `type` of a notification                               | URL         | [Notification Type](#notification-type)  | TRUE      |

##### Response parameters

``` http
HTTP/1.1 200 OK
```

##### Request example

```bash
curl -v -X POST \
-H "Accept: application/json" \
-H "Content-Type: application/json" \
-H "Authorization: SSWS ${api_token}" \
"https://${yourOktaDomain}/api/v1/roles/SUPER_ADMIN/subscriptions/IWA_AGENT/unsubscribe"
```

##### Response example

``` http
HTTP/1.1 200 OK
```

#### List subscriptions of a User

<ApiLifecycle access="ea" />

<ApiOperation method="get" url="/api/v1/users/${userId}/subscriptions" />

> List subscriptions of a User. We only support operations for current user right now.

Reviewer: what happens if this is called with a non-admin user id?

Author: If will first check the permission then valid the user id is the current user or not. So if the user doesn't have the permission, the call would fail for access denial

| notificationType | `type` of notification | URL         | [Notification Type](#notification-type)    | TRUE      |

##### Response parameters

[Subscription](#subscription-object)

##### Request example

```bash
curl -v -X GET \
-H "Accept: application/json" \
-H "Content-Type: application/json" \
-H "Authorization: SSWS ${api_token}" \
"https://${yourOktaDomain}/api/v1/users/00uuk0UVgUXjkIbPL0g3/subscriptions/IWA_AGENT"
```

##### Response example

```json
{
    "notificationType": "IWA_AGENT",
    "channels": [
        "email"
    ],
    "status": "subscribed",
    "_links": {
        "unsubscribe": {
            "href": "http://rain.okta1.com:1802/api/v1/users/00uuk0UVgUXjkIbPL0g3/subscriptions/IWA_AGENT/unsubscribe",
            "hints": {
                "allow": [
                    "POST"
                ]
            }
        },
        "self": {
            "href": "http://rain.okta1.com:1802/api/v1/users/00uuk0UVgUXjkIbPL0g3/subscriptions/IWA_AGENT"
        }
    }
}
```

#### Subscribe to a specific type

<ApiLifecycle access="ea" />

<ApiOperation method="post" url="/api/v1/users/${userId}/subscriptions/${notificationType}/subscribe" />

> Make a User subscribes to a specific type. We only support operations for current user right now. Requesting for other users will get AccessDeniedException.

Reviewer 2: ```suggestion
Subscribe a User to a specific type. We only support operations for current user right now. Requesting for other users will get AccessDeniedException.
```

##### Request parameters

| Parameter        | Description             | Param Type  | DataType                                 | Required  |
| :--------------- | :---------------------  | :---------- | :--------------------------------------- | :-------- |
| userId           | `id` of a User          | URL         | String                                   | TRUE      |
| notificationType | `type` of notification  | URL         | [Notification Type](#notification-type) | TRUE      |

##### Response parameters

``` http
HTTP/1.1 200 OK
```

##### Request example

```bash
curl -v -X POST \
-H "Accept: application/json" \
-H "Content-Type: application/json" \
-H "Authorization: SSWS ${api_token}" \
"https://${yourOktaDomain}/api/v1/users/00uuk0UVgUXjkIbPL0g3/subscriptions/IWA_AGENT/subscribe"
```

##### Response example

``` http
HTTP/1.1 200 OK
```

#### Unsubscribe to a specific type

<ApiLifecycle access="ea" />

<ApiOperation method="post" url="/api/v1/users/${userId}/subscriptions/${notificationType}/unsubscribe" />

> Make a User unsubscribes to a specific type. We only support operations for current user right now. Requesting for other users will get AccessDeniedException.

Reviewer 2: ```suggestion
Unsubscribe a User from a specific type. We only support operations for current user right now. Requesting for other users will get AccessDeniedException.
```

| :------------------ | :----------------------------------------------------------------- |
| `GROUP`             | Role was assigned to an admin Group of which the User is a member |
| `USER`              | Role was assigned to the User directly                             |

## Subscription object

### Examples

#### Sample Subscription

```json
{
    "notificationType": "IWA_AGENT",
    "channels": [
        "email"
    ],
    "status": "subscribed",
    "_links": {
        "unsubscribe": {
            "href": "http://rain.okta1.com:1802/api/v1/users/00uuk0UVgUXjkIbPL0g3/subscriptions/IWA_AGENT/unsubscribe",
            "hints": {
                "allow": [
                    "POST"
                ]
            }
        },
        "self": {
            "href": "http://rain.okta1.com:1802/api/v1/users/00uuk0UVgUXjkIbPL0g3/subscriptions/IWA_AGENT"
        }
    }
}
```

### Subscription properties

> The Subscription object defines several properties:

Author: `notificationType`  will always be valid in the response. If we request for invalid type in the reqquest, ApiValidationException will be thrown

| Property         | Description                                                  | DataType                                                        | Nullable   | Unique   | Read Only |
| :--------------- | :----------------------------------------------------------- | :-------------------------------------------------------------- | :--------- | :------- | :-------- |
| notificationType | Type of subscriptions                                        | [Notification Type](#notification-type)                         | FALSE      | TRUE     | TRUE      |

> | channels         | Array of sources that the user will get notification from    | array of [Channel](#subscription-channel)                       | FALSE      | FALSE    | TRUE      |

Reviewer: another note probably worth adding - or linking to the existing help doc - is that even if status shows subscribed, the admin may not get a notification because for example they may not have permission on the specific resource. For example, a help desk admin will only received user locked out emails for the users that are in groups to which the admin is scoped.

> | status           | Status of subscriptions                                      | `subscribed`, `unsubscribed`                                    | FALSE      | FALSE    | FALSE     |

Reviewer: I though we decided the lack of channel in channel array should be considered as whether admin is subscribed or not @yuanxue-okta to keep me honest. Not sure for example what does it mean to have more than one channel in the array with a single subscribed or unsubscribed status. I don't see how that will be consumable by the caller

Author: We don't depend on the `channel` any more since we introduced `status` in the subscription

| _links           | Discoverable resources related to the subscription           | [JSON HAL](http://tools.ietf.org/html/draft-kelly-json-hal-06)  | FALSE      | FALSE    | TRUE      |

### Subscription channel
We currently only support `email`

### Notification type

> Since we currently only support `email`, the notification types are the same as email types: `CONNECTOR_AGENT`, `USER_LOCKED_OUT`, `APP_IMPORT`, `LDAP_AGENT`, `AD_AGENT`, `OKTA_ANNOUNCEMENT`, `OKTA_ISSUE`, `OKTA_UPDATE`, `IWA_AGENT`, `USER_DEPROVISION`, `REPORT_SUSPICIOUS_ACTIVITY`, `RATELIMIT_NOTIFICATION`.

Author: How should we request for these?

> Since we currently only support `email`, the notification types are the same as email types: `CONNECTOR_AGENT`, `USER_LOCKED_OUT`, `APP_IMPORT`, `LDAP_AGENT`, `AD_AGENT`, `OKTA_ANNOUNCEMENT`, `OKTA_ISSUE`, `OKTA_UPDATE`, `IWA_AGENT`, `USER_DEPROVISION`, `REPORT_SUSPICIOUS_ACTIVITY`, `RATELIMIT_NOTIFICATION`.

Author: @bethwang-okta Do we want to add description for each type in the docs?

> Since we currently only support `email`, the notification types are the same as email types: `CONNECTOR_AGENT`, `USER_LOCKED_OUT`, `APP_IMPORT`, `LDAP_AGENT`, `AD_AGENT`, `OKTA_ANNOUNCEMENT`, `OKTA_ISSUE`, `OKTA_UPDATE`, `IWA_AGENT`, `USER_DEPROVISION`, `REPORT_SUSPICIOUS_ACTIVITY`, `RATELIMIT_NOTIFICATION`.

Reviewer 3: Here are descriptions for each email type:
`CONNECTOR_AGENT`System notification sent when an on-premises provisioning or Okta on-prem MFA agent disconnects or reconnects.
 `USER_LOCKED_OUT`System notification sent when a user is locked out from logging in to Okta.
`APP_IMPORT` System notification sent with the status of an app user import.
`LDAP_AGENT`System notification sent when an LDAP agent disconnects or reconnects.
`AD_AGENT` System notification sent when an AD agent disconnects or reconnects.
`OKTA_ANNOUNCEMENT` Okta communication sent for announcements and release notes.
`OKTA_ISSUE` Okta communication sent for trust incident alerts and updates
`OKTA_UPDATE` Okta communication sent for scheduled system updates
`IWA_AGENT`System notification sent when an IGA agent disconnects or reconnects.
`USER_DEPROVISION` System notification sent when a user is deprovisioned from apps.
`REPORT_SUSPICIOUS_ACTIVITY`System notification sent when a user reports suspicious activity.
`RATELIMIT_NOTIFICATION` System notification sent when an org reaches rate limit warning or violation thresholds.
